### ISSUE TITLE
don't barf on path_exists symlink

### ISSUE THREAD

**Issue by JeffAbrahamson** [i3/i3status#190]
I have the following directives in my `.i3status.conf`:

    order += "path_exists images"
    path_exists "images" {
	    path = "/home/jeff/images/.mounted"
	    format = "[images]"
	    format_down = ""
    }

In the event, images is a symlink:

    /home/jeff/images -> /home2/jeff

Similar directives (without symlinks) work fine, but this one causes i3status to fail when `/home2/` is not mounted.  The failure is that the i3 bar status changes from its usual useful information to

    Error: status_command process exited unexpectedly (exit 130)

I think the correct behaviour would be for path_exists to conclude that the path does not exist.

**Comment by stapelberg** [comment]
Which version of i3status is this? I cannot reproduce this issue. I added the following to my i3status config:

```
order += "path_exists images"
path_exists "images" {
    path = "/tmp/sym/foo"
    format = "images up"
    format_down = "images down"
}
```

I then created the symlink `/tmp/sym`: `ln -s /does/not/exist /tmp/sym`.

When running i3status on a terminal, I see “images down” in the output, which is expected.

Can you provide a backtrace when i3status crashes? See https://i3wm.org/docs/debugging-release-version.html for details

**Comment by JeffAbrahamson** [comment]
I think this is the only important part of the strace:

    write(2, "i3status: trying to auto-detect "..., 54i3status: trying to auto-detect output_format setting
    ) = 54
    ioctl(1, TCGETS, {B38400 opost isig icanon echo ...}) = 0
    write(2, "i3status: auto-detected \"term\"\n", 31i3status: auto-detected "term"
    ) = 31
    fstat(1, {st_mode=S_IFCHR|0620, st_rdev=makedev(136, 2), ...}) = 0
    socket(PF_INET, SOCK_DGRAM, IPPROTO_IP) = 3
    statfs("/", {f_type="EXT2_SUPER_MAGIC", f_bsize=4096, f_blocks=57204545, f_bfree=13607119, f_bavail=10695529, f_files=14540800, f_ffree=13715317, f_fsid={-1576804086, 1129614555}, f_namelen=255, f_frsize=4096, f_flags=4128}) = 0
    statfs("/home/jeff/images", 0x7fffec446590) = -1 ENOENT (No such file or directory)
    --- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0} ---
    +++ killed by SIGSEGV (core dumped) +++
    Segmentation fault (core dumped)

I've attached the entire strace (zipped, because that's what github wants).
[i3status.strace.zip](https://github.com/i3/i3status/files/702775/i3status.strace.zip)

**Comment by JeffAbrahamson** [comment]
Oh, and

    jeff@starshine:tmp $ i3status --version
    i3status 2.10 (2016-01-01) © 2008 Michael Stapelberg and contributors
    jeff@starshine:tmp $

**Comment by JeffAbrahamson** [comment]
The i3status binary is stripped.  The i3-wm-dbg package (apparently) doesn't provide debug symbols for it, so gdb isn't getting a meaningful backtrace out of it.  Is there anything I can do short of building from git?  Which, presumably, I can do, of course, but I'm always open to the lazy approach if it's there.

**Comment by stapelberg** [comment]
You’d need to install i3status-dbgsym (not i3-wm-dbg), but that doesn’t seem to be built for amd64 yet. I’d recommend to build from git after installing the dependencies via `apt-get build-dep i3status`

**Comment by stapelberg** [comment]
The `statfs` call from your strace output is done by the disk_info module, not path_exists.

Hence, I believe this is a duplicate of https://github.com/i3/i3status/issues/92

We should do a new release to get the fix out to people.

**Comment by JeffAbrahamson** [comment]
Fwiw, I ran `git bisect` and learned this:

The format_down flag has existed for quite some time (at least since f947d0a4 on 2009-10-11), but in this context it first works here:

    commit 5c094a54932edebf6e44c4a7119fa510ddf7ae7a
    Author: Christian Kohlstedde <christian@kohlsted.de>
    Date:   Thu Apr 2 22:31:06 2015 +0200

        Adding the option "format_down" to path_exists and run_watch.
        This fixes #1.

The symlink crash bug has not always existed.  It was introduced here:

    commit 75a835742e60311f7f27bf44af862caad67e3a23
    Author: Ingo Bürk <ingo.buerk@tngtech.com>
    Date:   Fri May 29 08:41:11 2015 +0200

        Use format_not_mounted if the directory for a disk_info block does not exist.
        This is useful for dynamically mounted devices like thumb drives that get mounted into a directory such as
        '/media/$UID/$NAME'.

        fixes #33

That change looks like this (`git diff d1aa135d2801d361f2eccc5efab08a6385ecb440..75a835742e60311f7f27bf44af862caad67e3a23`):

    diff --git a/src/print_disk_info.c b/src/print_disk_info.c
    index 69d7b8c..7fd47b9 100644
    --- a/src/print_disk_info.c
    +++ b/src/print_disk_info.c
    @@ -124,10 +124,11 @@ void print_disk_info(yajl_gen json_gen, char *buffer, const char *path, const ch
     #else
         struct statvfs buf;

    -    if (statvfs(path, &buf) == -1)
    -        return;
    -
    -    if (format_not_mounted != NULL) {
    +    if (statvfs(path, &buf) == -1) {
    +        /* If statvfs errors, e.g., due to the path not existing,
    +         * we use the format for a not mounted device. */
    +        format = format_not_mounted;
    +    } else if (format_not_mounted != NULL) {
             FILE *mntentfile = setmntent("/etc/mtab", "r");
             struct mntent *m;
             bool found = false;

It was fixed here (shortly after the current release):

    commit fad9c8237c63290aa9adde5ba6ffa6759f578645
    Author: Michael Stapelberg <michael@stapelberg.de>
    Date:   Tue Jan 12 09:49:59 2016 +0100

        Bugfix: check that format_not_mounted is not NULL before using it

        This issue was introduced with commit 75a83574.

        fixes #92

That change looks like this (`git diff 617953117796d8ae0cff11a4192a923bdce34452..fad9c8237c63290aa9adde5ba6ffa6759f578645`):

    @@ -127,7 +127,9 @@ void print_disk_info(yajl_gen json_gen, char *buffer, const char *path, const ch
         if (statvfs(path, &buf) == -1) {
             /* If statvfs errors, e.g., due to the path not existing,
              * we use the format for a not mounted device. */
    -        format = format_not_mounted;
    +        if (format_not_mounted != NULL) {
    +            format = format_not_mounted;
    +        }
         } else if (format_not_mounted != NULL) {
             FILE *mntentfile = setmntent("/etc/mtab", "r");
             struct mntent *m;
